// @dodona/dolos-lib/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DefaultMap = exports.PairedOccurrence = exports.SharedFingerprint = exports.SemanticAnalyzer = exports.EmptyTokenizer = exports.Dolos = exports.Index = exports.TokenizedFile = exports.File = exports.Options = exports.Pair = exports.Region = exports.Fragment = exports.Report = void 0;
var report_1 = require("./lib/analyze/report");
Object.defineProperty(exports, "Report", { enumerable: true, get: function () { return report_1.Report; } });
var fragment_1 = require("./lib/analyze/fragment");
Object.defineProperty(exports, "Fragment", { enumerable: true, get: function () { return fragment_1.Fragment; } });
var region_1 = require("./lib/util/region");
Object.defineProperty(exports, "Region", { enumerable: true, get: function () { return region_1.Region; } });
var pair_1 = require("./lib/analyze/pair");
Object.defineProperty(exports, "Pair", { enumerable: true, get: function () { return pair_1.Pair; } });
var options_1 = require("./lib/util/options");
Object.defineProperty(exports, "Options", { enumerable: true, get: function () { return options_1.Options; } });
var file_1 = require("./lib/file/file");
Object.defineProperty(exports, "File", { enumerable: true, get: function () { return file_1.File; } });
var tokenizedFile_1 = require("./lib/file/tokenizedFile");
Object.defineProperty(exports, "TokenizedFile", { enumerable: true, get: function () { return tokenizedFile_1.TokenizedFile; } });
var index_1 = require("./lib/analyze/index");
Object.defineProperty(exports, "Index", { enumerable: true, get: function () { return index_1.Index; } });
var dolos_1 = require("./dolos");
Object.defineProperty(exports, "Dolos", { enumerable: true, get: function () { return dolos_1.Dolos; } });
var emptyTokenizer_1 = require("./lib/tokenizer/emptyTokenizer");
Object.defineProperty(exports, "EmptyTokenizer", { enumerable: true, get: function () { return emptyTokenizer_1.EmptyTokenizer; } });
var SemanticAnalyzer_1 = require("./lib/analyze/SemanticAnalyzer");
Object.defineProperty(exports, "SemanticAnalyzer", { enumerable: true, get: function () { return SemanticAnalyzer_1.SemanticAnalyzer; } });
var sharedFingerprint_1 = require("./lib/analyze/sharedFingerprint");
Object.defineProperty(exports, "SharedFingerprint", { enumerable: true, get: function () { return sharedFingerprint_1.SharedFingerprint; } });
var pairedOccurrence_1 = require("./lib/analyze/pairedOccurrence");
Object.defineProperty(exports, "PairedOccurrence", { enumerable: true, get: function () { return pairedOccurrence_1.PairedOccurrence; } });
var defaultMap_1 = require("./lib/util/defaultMap");
Object.defineProperty(exports, "DefaultMap", { enumerable: true, get: function () { return defaultMap_1.DefaultMap; } });
__exportStar(require("./lib/util/utils"), exports);
