// @dodona/dolos-lib/dist/lib/hashing/noFilter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NoFilter = void 0;
const hashFilter_1 = require("./hashFilter");
const rollingHash_1 = require("./rollingHash");
class NoFilter extends hashFilter_1.HashFilter {
    /**
     * Generates a HashFilter object with given kgram size. It will not hashing
     * anything and return all hashes
     *
     * @param k The kgram size of which hashes are calculated
     * @param debug Whether to output debugging information in fingerprints.
     */
    constructor(k, debug = false) {
        super(debug);
        this.k = k;
    }
    /**
     * Returns an async iterator that yields tuples containing a hashing and its
     * corresponding kgram position. Can be called successively on multiple files.
     *
     * @param tokens The list of tokens to process.
     */
    async *fingerprints(tokens) {
        const hash = new rollingHash_1.RollingHash(this.k);
        let window = [];
        let filePos = -1 * this.k;
        for await (const [byte, token] of this.hashTokens(tokens)) {
            filePos++;
            window = window.slice(-this.k + 1);
            window.push(token);
            if (filePos < 0) {
                hash.nextHash(byte);
                continue;
            }
            yield {
                hash: hash.nextHash(byte),
                start: filePos,
                stop: filePos + this.k - 1,
                data: this.kgramData ? window : null,
            };
        }
    }
}
exports.NoFilter = NoFilter;
