// @dodona/dolos-lib/dist/lib/analyze/sharedFingerprint.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SharedFingerprint = void 0;
const defaultMap_1 = require("../util/defaultMap");
const identifiable_1 = __importDefault(require("../util/identifiable"));
class SharedFingerprint extends identifiable_1.default {
    constructor(hash, kgram) {
        super();
        this.hash = hash;
        this.kgram = kgram;
        this.partMap = new defaultMap_1.DefaultMap(() => new Set());
    }
    add(part) {
        this.partMap.get(part.file).add(part);
    }
    addAll(parts) {
        parts.forEach(p => this.add(p));
    }
    parts() {
        return Array.of(...this.partMap.values())
            .map(set => Array.of(...set))
            .flat();
    }
    files() {
        return Array.of(...new Set(this.parts().map(p => p.file)));
    }
}
exports.SharedFingerprint = SharedFingerprint;
